calicoctl/commands/ipam/show.go

```diff
@@ -21,83 +21,135 @@ import (
-func Show(args []string) error {
-	doc := constants.DatastoreIntro + `Usage:
-  calicoctl ipam show [--ip=<IP> | --show-blocks] [--config=<CONFIG>]
-
-Options:
-  -h --help             Show this screen.
-     --ip=<IP>          Report whether this specific IP address is in use.
-     --show-blocks      Show detailed information for IP blocks as well as pools.
-  -c --config=<CONFIG>  Path to the file containing connection configuration in
-                        YAML or JSON format.
-                        [default: ` + constants.DefaultConfigPath + `]
+func getPoolDetails(ctx context.Context, ippoolClient clientv3.IPPoolInterface, bc bapi.Client) ([]*ipam.PoolDetails, error) {
+	var details []*ipam.PoolDetails
 
-Description:
-  The ipam show command prints information about a given IP address, or about
-  overall IP usage.
-`
-	parsedArgs, err := docopt.Parse(doc, args, true, "", false, false)
+	allPools, err := ippoolClient.List(ctx, options.ListOptions{})
 	if err != nil {
-		return fmt.Errorf("Invalid option: 'calicoctl %s'. Use flag '--help' to read about a specific subcommand.", strings.Join(args, " "))
+		return nil, err
 	}
-	if len(parsedArgs) == 0 {
-		return nil
+
+	for _, pool := range allPools.Items {
+		details = append(details, &ipam.PoolDetails{
+			Name: pool.Name,
+			CIDR: net.MustParseNetwork(pool.Spec.CIDR).IPNet,
+		})
 	}
 
-	ctx := context.Background()
+	// Read details for all blocks.
+	blocks, err := bc.List(ctx, model.BlockListOptions{}, "")
+	if err != nil {
+		return nil, err
+	}
+	for _, kvp := range blocks.KVPairs {
+		b := kvp.Value.(*model.AllocationBlock)
+
+		// Find which pool this block belongs to.
+		for _, poolDetails := range details {
+			if b.CIDR.IsNetOverlap(poolDetails.CIDR) {
+				log.Debugf("Block CIDR %v belongs to pool %v", b.CIDR, poolDetails.Name)
+
+				blockDetail := &ipam.BlockDetail{CIDR: b.CIDR.IPNet}
+				if b.Affinity != nil {
+					blockDetail.Node = *b.Affinity
+				}
+
+				for _, attributes := range b.Attributes {
+					blockDetail.Consumers = append(blockDetail.Consumers,
+						ipam.Consumer{Name: attributes.AttrSecondary[model.IPAMBlockAttributePod],
+							Namespace: attributes.AttrSecondary[model.IPAMBlockAttributeNamespace],
+							Node:      attributes.AttrSecondary[model.IPAMBlockAttributeNode]})
+				}
+
+				poolDetails.BlockDetails = append(poolDetails.BlockDetails, blockDetail)
+				break
+			}
+		}
+	}
+	return details, nil
+}
 
-	// Create a new backend client from env vars.
-	cf := parsedArgs["--config"].(string)
-	client, err := clientmgr.NewClient(cf)
+func showBorrowedDetails(ctx context.Context, ippoolClient clientv3.IPPoolInterface, bc bapi.Client) error {
+	details, err := getPoolDetails(ctx, ippoolClient, bc)
 	if err != nil {
 		return err
 	}
+	table := tablewriter.NewWriter(os.Stdout)
+	table.SetHeader([]string{"POOL", "BLOCK", "NODE OWNING BLOCK", "POD", "NAMESPACE", "NODE BORROWING IP"})
```

Comment:
I think we need to include the IP address itself in this output. And actually, I'm not sure if we need to show the pool at all?

I was imagining something like this:

```
IP                   BORROWING-NODE         BLOCK               BLOCK OWNER      TYPE    ALLOCATED-TO
192.168.0.1          my-node-1              192.168.0.0/26      my-node-2        pod     kube-system/coredns-pod-2
192.168.0.2          my-node-4              192.168.0.0/26      my-node-2        vxlan   my-node-4
192.168.0.3          my-node-8              192.168.0.0/26      my-node-2        N/A     N/A
 
```

This shows the three types of IP allocations we might have:
- Pod allocation
- tunnel address allocation (in this case VXLAN, but it could also be IPIP)
- unknown - i.e., doesn't have attributes attached to it

Reply:
That's much better. Thank you!